### PR TITLE
Late cast expansion: more improvements #2

### PR DESCRIPTION
**Pull request by EgorBo**
Almost there to remove cast expansion from importer. The only two missing pieces are redundant nullchecks and I completely forgout about cast expansion with runtime lookups (where cast's argument not a class handle, but GT_RUNTIMELOOKUP).

### CODE REVIEW

**Comment by EgorBo** [comment]
@MihuBot

**Comment by EgorBo** [comment]
@jakobbotsch @dotnet/jit-contrib PTAL - it moved more logic to the late phase and fixed a few issues.

The last two missing pieces before I remove the early expansions are:
1) Redundant nullchecks (to be handled in the other PR)
2) I totally forgot about cast expansions against runtime lookups, e.g.

```cs
[MethodImpl(MethodImplOptions.NoInlining)]  
static bool Foo<T>(object o)
{
    return o is Foo<T>;
}

public sealed class Foo<T>
{

}
```

Also https://github.com/dotnet/source-build/issues/4007 bug doesn't repro with this PR + https://github.com/dotnet/runtime/pull/97387

File: src/coreclr/jit/helperexpansion.cpp [hunk 1]
```diff
@@ -2268,7 +2279,18 @@ bool Compiler::fgLateCastExpansionForCall(BasicBlock** pBlock, Statement* stmt,
...
+        GenTreeCall* helperCall = call;
+        if (typeCheckFailedAction == TypeCheckFailedAction::CallHelper_Specialized)
+        {
+            // Re-create the call with a specialized helper
+            // For JIT we probably can just change the gtCallMethHnd in the existing call, but let's unify with AOT
+            helperCall = gtNewHelperCallNode(CORINFO_HELP_CHKCASTCLASS_SPECIAL, call->TypeGet(),
+                                             helperCall->gtArgs.GetUserArgByIndex(0)->GetNode(),
+                                             helperCall->gtArgs.GetUserArgByIndex(1)->GetNode());
+            fgMorphCall(helperCall);
+            gtSetEvalOrder(helperCall);
```
Comment (jakobbotsch) [inline src/coreclr/jit/helperexpansion.cpp:2291]: Just curious -- how is NAOT and JIT different here?

Comment (SingleAccretion) [reply src/coreclr/jit/helperexpansion.cpp:2291]: AOT needs you to set the entry point. This could be done by mutating the node, theoretically.

**Review by jakobbotsch — APPROVED** [review]

File: src/coreclr/jit/helperexpansion.cpp (hunk 1 above)
Comment (jakobbotsch) [reply src/coreclr/jit/helperexpansion.cpp:2291]: Hmm, but what is setting the entry point here?

Comment (SingleAccretion) [reply src/coreclr/jit/helperexpansion.cpp:2291]: Good point - nothing, so that is not the reason then.

Comment (EgorBo) [reply src/coreclr/jit/helperexpansion.cpp]: Ah, it's not needed at all - `CORINFO_HELP_CHKCASTCLASS` is not used for AOT anyway so we're only improving JIT here (copies importer's behavior)

Comment (EgorBo) [reply src/coreclr/jit/helperexpansion.cpp]: Ah, looks like it's needed if I were changing entrypoint for CORINFO_HELP_READYTORUN_CHKCAST

**Pull request merged**